orc.js

```diff
@@ -1,42 +1,87 @@
-var app = require('commander');
+var commander = require('commander');
 var git = require('gitty');
 var path = require('path');
-
-// process.stdin.resume();
-// process.stdin.setEncoding('utf8');
-// process.stdin.on('data', function(data) {
-//  process.stdout.write(data);
-// });
+var pushover   = require('pushover');
 
 process.on('SIGINT', function () {
   console.log('Got a SIGINT. Goodbye cruel world');
   process.exit(0);
 });
 
-function _requireCommit() {
-
+function _setup(dir) {
+  // TODO: march up to root dir just to make output consistent
+  var repo = git(dir);
+  branches = repo.getBranchesSync();
+  if (!branches.current) {
+    throw new Error('error: '+repo.name+' doesn\'t appear to be a valid git repo');
+  }
+  return { repo_dir: dir, repo: git(dir)};
 }
 
-function checkpoint() {
+function _requireClean(orc) {
+  var status = orc.repo.statusSync();
+  if(status.unstaged.length > 0) {
+    throw new Error('unstaged files, commit before running this.');
+  }
+  if(status.untracked.length > 0) {
+    throw new Error('untracked files, add or ignore before running this.');
+  }
+}
 
+// commit local changes
+// push to remote
+function checkpoint(orc, cb) {
+  if(orc.repo.statusSync().untracked.length > 0) {
+    throw new Error('untracked files, add or ignore before running this.');
+  }
+  if(!orc.repo.commitSync('ORC-CHECKPOINT')) {
+    throw new Error('checkpoint commit failed');
+  }
+  orc.repo.push('origin', 'master', function(err, result) {
+    if (err) {
+      cb('error pushing checkpoint: '+JSON.stringify(err));
+      return;
+    }
+    cb(null, true);
+  });
 }
 
-function pull() {
-	// TODO
+function pull(orc, cb) {
+  // TODO
 }
 
-function push() {
-	// TODO
+function push(orc, cb) {
+  // TODO
 }
 
-function branch() {
-	// TODO
+function createBranch(orc, branchName, cb) {
+  _requireClean(orc);
+  orc.repo.checkoutSync('master');
+  orc.repo.pull('origin','master', function(err, res) {
+    if (err) {
+      throw new Error('failed to pull from master: '+JSON.stringify(err));
+    }
+    if (orc.repo.createBranchSync(branchName) !== '') {
+      throw new Error('failed to create branch '+branchName);
+    }
+    if(!orc.repo.checkoutSync(branchName)) {
+      throw new Error('couldn\'t checkout branch '+branchName);
+    }
+    orc.repo.push('origin', branchName, ['--set-upstream'], function(err, res) {
+      if (err) {
+        cb(err);
+      } else {
+        cb(null, true);
+      }
+    });
+  });
+  // TODO: block until done
```

Comment:
initial version of this was all synchronous, as @conordickinson and I talked about yesterday this eventually bites you in the ass. switching to all-cb version